Libraries/MLXLMCommon/Evaluate.swift

```diff
@@ -0,0 +1,556 @@
+        """
+    }
+}
+
+/// Action from token visitor callback in ``generate(input:parameters:context:didGenerate:)``.
+public enum GenerateDisposition: Sendable {
+    /// keep producing tokens until an EOS token is produced
+    case more
+
+    /// stop producing tokens, e.g. a token limit has been hit
+    case stop
+}
+
+/// Given prompt tokens generate text using the given model and parameters.
+///
+/// ``generate(input:parameters:context:didGenerate:)`` is the preferred call.
+///
+/// - Parameters:
+///   - promptTokens: tokenized prompt
+///   - parameters: generation parameters
+///   - model: model to evaluate
+///   - tokenizer: tokenizer to convert tokens back into strings and recognizer special tokens
+///   - extraEOSTokens: any additional stop tokens
+///   - didGenerate: visitor for the tokens as they are generated
+@available(*, deprecated, message: "please use generate(input:parameters:context:didGenerate:)")
+public func generate(
+    promptTokens: [Int], parameters: GenerateParameters, model: any LanguageModel,
+    tokenizer: Tokenizer,
+    extraEOSTokens: Set<String>? = nil,
+    didGenerate: ([Int]) -> GenerateDisposition
+) throws -> GenerateResult {
+    let tokens = MLXArray(promptTokens)
+    let iterator = try TokenIterator(
+        prompt: tokens, model: model, parameters: parameters)
+
+    // this is a compatibility cover -- create the required values
+    // for the iteration
+    let input = LMInput(tokens: tokens)
+    let configuration = ModelConfiguration(id: "stand-in", extraEOSTokens: extraEOSTokens ?? [])
+    let context = ModelContext(
+        configuration: configuration, model: model, processor: StandInUserInputProcessor(),
+        tokenizer: tokenizer)
+
+    return generate(
+        input: input, context: context, iterator: iterator, didGenerate: didGenerate)
+}
+
+/// Generate tokens from an ``LMInput`` and a ``ModelContext``.
+///
+/// For example:
+///
+/// ```swift
+/// let generateParameters: GenerateParameters
+/// let input: UserInput
+/// let context: ModelContext
+///
+/// let lmInput = try context.processor.prepare(input: input)
+/// let result = generate(input: lmInput,
+///     parameters: generateParameters,
+///     context: context) { tokens in
+///     .more
+/// }
+/// ```
+///
+/// Internally this constructs a ``TokenIterator`` and calls
+/// ``generate(input:context:iterator:didGenerate:)``
+///
+/// - Parameters:
+///   - input: prepared language model input
+///   - parameters: parameters controlling the token generation
+///   - context: model context (model and tokenizer)
+///   - didGenerate: token visitor that can output tokens as they are generated and indicate early stop
+/// - Returns: the generated output
+public func generate(
+    input: LMInput, parameters: GenerateParameters, context: ModelContext,
+    didGenerate: ([Int]) -> GenerateDisposition
+) throws -> GenerateResult {
```

Comment:
This is the new interface to `generate()` -- simplified with the addition of `LMInput` and `ModelContext`.  Minimally token generation looks like this:

```swift
let parameters = GenerateParameters()
let input = UserInput(prompt: "tell me a story")

let result = try await modelContainer.perform { [generate, input] context in
    let input = try await context.processor.prepare(input: input)
    return try generate(input: input, parameters: parameters, context: context) { token in
        // print tokens as they are generated, stop early, etc.
        return .more
    }
}
```